### PR TITLE
feat: add email changing

### PR DESCRIPTION
**Pull request by karolsojko**
Internal link: https://app.asana.com/0/1199914526147392/1200887853314814/f

### CODE REVIEW

File: packages/snjs/lib/services/credential_service.ts [hunk 1]
```diff
@@ -518,4 +445,103 @@ export class SNCredentialService extends PureService<AccountEvent> {
...
+        await this.syncService.sync({ awaitAll: true });
+
+        return { error: Error(CredentialsChangeStrings.Failed) };
+      }
+    }
+
+    return result.response;
+  }
+
+  private async recomputeRootKeys(parameters: {
```
Comment (mobitar) [inline packages/snjs/lib/services/credential_service.ts:525]: Should be made more specific that this is just for the change function.

```suggestion
  private async recomputeRootKeysForCredentialChange(parameters: {
```

File: packages/snjs/lib/services/credential_service.ts [hunk 2]
```diff
@@ -518,4 +445,103 @@ export class SNCredentialService extends PureService<AccountEvent> {
...
+    });
+    this.unlockSyncing();
+    if (!result.response.error) {
+      const rollback = await this.protocolService.createNewItemsKeyWithRollback();
+      await this.protocolService.reencryptItemsKeys();
+      await this.syncService.sync({ awaitAll: true });
+      const defaultItemsKey = this.protocolService.getDefaultItemsKey() as SNItemsKey;
+      const itemsKeyWasSynced = !defaultItemsKey.neverSynced;
+      if (!itemsKeyWasSynced) {
+        const emailShouldBeRolledBack = parameters.newEmail !== undefined && parameters.newEmail !== currentEmail
```
Comment (mobitar) [inline packages/snjs/lib/services/credential_service.ts:507]: A few missing semicolons in file, better to lint entire file and commit

```suggestion
        const emailShouldBeRolledBack = parameters.newEmail !== undefined && parameters.newEmail !== currentEmail;
```

**Review by mobitar — APPROVED** [review]

File: packages/snjs/lib/services/credential_service.ts (hunk 2 above)
Comment (karolsojko) [reply packages/snjs/lib/services/credential_service.ts:507]: 👍 

will fix though linter doesn't show any warnings nor errors hmm :/

**Review by mobitar — APPROVED** [review]

File: packages/snjs/test/lib/session.test.js [hunk 3]
```diff
@@ -164,6 +164,103 @@ describe('server session', function () {
...
+  it('change email request should be successful with a valid access token', async function () {
+    let { application, password } = await Factory.createAndInitSimpleAppContext({
+      registerUser: true
+    });
+    const newEmail = Uuid.GenerateUuidSynchronously();
+    const changeEmailResponse = await application.changeEmail(
+      newEmail,
+      password
+    );
+    console.log(changeEmailResponse.data)
```
Comment (johnny243) [inline packages/snjs/test/lib/session.test.js:176]: Debugging here? 😄

File: packages/snjs/test/lib/session.test.js [hunk 4]
```diff
@@ -164,6 +164,103 @@ describe('server session', function () {
...
+
+    application = await Factory.signOutApplicationAndReturnNew(
+      application
+    );
+    const loginResponse = await Factory.loginToApplication({
+      application: application,
+      email: newEmail,
+      password: password,
+    });
+    console.log(loginResponse)
```
Comment (johnny243) [inline packages/snjs/test/lib/session.test.js:189]: Same here

File: packages/snjs/test/lib/session.test.js (hunk 3 above)
Comment (karolsojko) [reply packages/snjs/test/lib/session.test.js:176]: Yup :P

File: packages/snjs/test/lib/session.test.js (hunk 4 above)
Comment (karolsojko) [reply packages/snjs/test/lib/session.test.js:189]: indeed :)

**Pull request merged**